libcxx/src/new.cpp

```diff
@@ -38,39 +44,53 @@ static void* operator_new_impl(std::size_t size) noexcept {
 
 _LIBCPP_WEAK void* operator new(std::size_t size) _THROW_BAD_ALLOC {
   void* p = operator_new_impl(size);
-#  ifndef _LIBCPP_HAS_NO_EXCEPTIONS
   if (p == nullptr)
-    throw std::bad_alloc();
-#  endif
+    __throw_bad_alloc_shim();
   return p;
 }
 
 _LIBCPP_WEAK void* operator new(size_t size, const std::nothrow_t&) noexcept {
+#  ifdef _LIBCPP_HAS_NO_EXCEPTIONS
+  _LIBCPP_ASSERT_SHIM(
```

Comment:
@var-const This is a rare example of a case where I think it makes sense to unconditionally have a `_LIBCPP_ASSERT` -- this should be enabled regardless of the hardening mode IMO. It is similar to when we call `std::__throw_logic_error()` and friends with `-fno-exceptions` -- that aborts the program regardless of the hardening mode.

Reply:
I think you're forgetting how hot of a path this is. I don't think we want to be doing anything more than we need in the majority of cases.

Reply:
We could make this part of the `DEBUG_LITE` mode then, I guess.